### PR TITLE
feat: Add code

### PR DESCRIPTION
**Pull request by averikitsch**
Will follow up on automating NPM publishing

### CODE REVIEW

**Comment by averikitsch** [comment]
I'm not sure why component installation works when this library is embedded in an action versus these tests

**Comment by bharathkkb** [comment]
@averikitsch I tried these locally and beta seemed to pass for me. For gsutil I had to bump the gcloud version, maybe an issue with json parsing in that specific version?

**Comment by averikitsch** [comment]
Code is ready for review. I will do another review of the documentation and workflows, then plan for publishing to NPM.

File: src/index.ts [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+ */
+export function isInstalled(version?: string): boolean {
+  let toolPath;
+  if (version) {
+    toolPath = toolCache.find('gcloud', version);
+    return toolPath != undefined && toolPath !== '';
+  } else {
+    toolPath = toolCache.findAllVersions('gcloud');
+    return toolPath.length > 0;
+  }
```
Comment (bharathkkb) [inline src/index.ts:44]: nit: redundant else
```suggestion
toolPath = toolCache.findAllVersions('gcloud');
return toolPath.length > 0;
```

File: src/index.ts [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+    output += data.toString();
+  };
+  const options = {
+    listeners: {
+      stdout,
+      stderr,
+    },
+    silent,
+  };
+
```
Comment (bharathkkb) [inline src/index.ts:82]: Looks like there a new [execOutput](https://github.com/actions/toolkit/blob/f0b00fd201c7ddf14e1572a10d5fb4577c4bd6a2/packages/exec/src/interfaces.ts#L41) which may simplify this and others

File: src/index.ts [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+export async function setProjectWithKey(
+  serviceAccountKey: string,
+): Promise<string> {
+  const serviceAccountJson = parseServiceAccountKey(serviceAccountKey);
+  await setProject(serviceAccountJson.project_id);
+  return serviceAccountJson.project_id;
+}
+
+/**
+ * Sets the GCP Project Id in the gcloud config.
```
Comment (bharathkkb) [inline src/index.ts:249]: nit: add installComponent description

File: src/index.ts [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+ *
+ * @param component - gcloud component group to install ie alpha, beta.
+ * @returns CMD output
+ */
+export async function installComponent(
```
Comment (bharathkkb) [inline src/index.ts:254]: Why not support a list of components?
```
gcloud components install COMPONENT-1 COMPONENT-2
```

File: src/index.ts [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+      toolCommand,
+      ['--quiet', 'components', 'install', component],
+      options,
+    );
+  } catch (err) {
+    throw new Error(`Unable to install component: ${component}`);
+  }
+}
+
+export async function runCmdWithJsonFormat(
```
Comment (bharathkkb) [inline src/index.ts:273]: nit: description

File: tests/download-util.test.ts [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+      await io.rmRF(toolDir);
+      await io.rmRF(tempDir);
+    } catch (err) {
+      console.error('Error occurred during test cleanup: ' + err);
+    }
+  });
+
+  it('downloads and extracts linux version', async function() {
+    if (os.platform() === 'win32') {
+      // https://github.com/actions/toolkit/issues/194
```
Comment (bharathkkb) [inline tests/download-util.test.ts:52]: this seems closed, might be worth running a test without this guard to confirm
https://github.com/actions/toolkit/issues/194

File: tests/download-util.test.ts [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+    const url = await getReleaseURL('darwin', 'x86_64', TEST_SDK_VERSION);
+    const extPath = await downloadUtil.downloadAndExtractTool(url);
+    expect(extPath).to.be;
+    expect(fs.existsSync(extPath)).to.be.true;
+  });
+
+  it('errors on download not found', async function() {
+    downloadUtil
+      .downloadAndExtractTool('fakeUrl')
```
Comment (bharathkkb) [inline tests/download-util.test.ts:85]: Would it be better to compare real error message?
```ts
downloadUtil
      .downloadAndExtractTool('fakeUrl')
         .should.eventually.be.rejectedWith("unable to find url")
```

File: tests/format-url.test.ts [hunk 8]
```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright 2019 Google LLC
```
Comment (bharathkkb) [inline tests/format-url.test.ts:2]: nit: year

File: tests/download-util.test.ts [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+ * Tests download-util.
+ */
+// import 'mocha';
```
Comment (bharathkkb) [inline tests/download-util.test.ts:20]: Should this be uncommented?

File: tests/index.test.ts [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+  afterEach(async function() {
+    try {
+      await io.rmRF(toolDir);
+      await io.rmRF(tempDir);
+    } catch (err) {
+      console.error('Error occurred during test cleanup: ' + err);
+    }
+  });
+
+  let version = '349.0.0';
```
Comment (bharathkkb) [inline tests/index.test.ts:41]: Why not use `TEST_SDK_VERSION`?

File: tests/index.test.ts [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+  });
+
+  it('returns false if not installed', function() {
+    const installed = setupCloudSDK.isInstalled();
+
+    expect(installed).eql(false);
+  });
+
+  it('returns false if version is not installed', function() {
+    const installed = setupCloudSDK.isInstalled(version);
```
Comment (bharathkkb) [inline tests/index.test.ts:56]: To test this should we install TEST_SDK_VERSIONS[-1] and check isInstalled returns false for TEST_SDK_VERSIONS[-2] ?

**Review by bharathkkb — APPROVED** [review]
Overall LGTM
Some small nits below

File: docs/README.md [hunk 12]
```diff
@@ -0,0 +1,66 @@
+@google-github-actions/setup-cloud-sdk / [Exports](modules.md)
+
+# [Google GitHub actions: Setup Cloud SDK Client](https://github.com/google-github-actions)
```
[the remark this answers is not in the document]
Comment (averikitsch) [reply docs/README.md:3]: I am generating docs and it copies the README into the docs directory

File: docs/modules.md [hunk 13]
```diff
@@ -0,0 +1,294 @@
+[@google-github-actions/setup-cloud-sdk](README.md) / Exports
+
+# @google-github-actions/setup-cloud-sdk
+
+## Table of contents
+
+### Functions
```
[the remark this answers is not in the document]
Comment (averikitsch) [reply docs/modules.md:7]: I am generating this using typedoc (`npm run docs`)

File: src/index.ts (hunk 4 above)
Comment (averikitsch) [reply src/index.ts:254]: updated.

File: src/index.ts (hunk 2 above)
Comment (averikitsch) [reply src/index.ts:82]: It is somewhat undetermined where gcloud will print to (stdout and stderr). This way keeps it easier to combine both outputs.

Comment (bharathkkb) [reply src/index.ts:82]: makes sense

**Pull request merged**